dipy/align/imaffine.py

```diff
@@ -0,0 +1,1044 @@
+        (`moving` grid) and interpolates the `moving` image at that point
+        to obtain the intensity value to be placed at `x` in the resulting
+        grid. The `transform_inverse` method performs the opposite operation
+        mapping points in the codomain to points in the domain.
+
+    ParzenJointHistogram: computes the marginal and joint distributions of
+        intensities of a pair of images, using Parzen windows [Parzen62]
+        with a cubic spline kernel, as proposed by Mattes et al. [Mattes03].
+        It also computes the gradient of the joint histogram w.r.t. the
+        parameters of a given transform.
+
+    ParzenMutualInformation: inherits from `ParzenJointHistogram` and
+        computes the Mutual Information similarity measure and its gradient
+        from the joint and marginal distributions of paired intensities,
+        which are computed by the super class.
+
+    MIMetric: it is a Python interface to the ParzenMutualInformation cython
+        class, which is used to compute the value and gradient the way
+        `Optimizer` needs them. That is, given a set of transform parameters,
+        it will use `ParzenMutualInformation` methods to compute the value and
+        gradient evaluated at the given parameters.
+
+    AffineRegistration: it runs the multi-resolution registration, putting
+        all the pieces together. It needs to create the scale space of the
+        images and run the multi-resolution registration by using the Metric
+        and the Optimizer at each level of the Gaussian pyramid. At each
+        level, it will setup the metric to compute value and gradient of the
+        metric with the input images with different levels of smoothing.
+
+    References
+    ----------
+    [Parzen62] E. Parzen. On the estimation of a probability density
+               function and the mode. Annals of Mathematical Statistics,
+               33(3), 1065-1076, 1962.
+    [Mattes03] Mattes, D., Haynor, D. R., Vesselle, H., Lewellen, T. K.,
+               & Eubank, W. PET-CT image registration in the chest using
+               free-form deformations. IEEE Transactions on Medical
+               Imaging, 22(1), 120-8, 2003.
+"""
+
+import numpy as np
+import numpy.linalg as npl
+import scipy.ndimage as ndimage
+from ..core.optimize import Optimizer
+from ..core.optimize import SCIPY_LESS_0_12
+from . import vector_fields as vf
+from . import VerbosityLevels
+from .parzenhist import ParzenMutualInformation, sample_domain_regular
+from .imwarp import (get_direction_and_spacings, ScaleSpace)
+from .scalespace import IsotropicScaleSpace
+
+
+class AffineMap(object):
+    def __init__(self, affine, domain_grid_shape=None, domain_grid2world=None,
+                 codomain_grid_shape=None, codomain_grid2world=None):
+        """ AffineMap
+
+        Implements an affine transformation whose domain is given by
+        `domain_grid` and `domain_grid2world`, and whose co-domain is
```

Comment:
If I understand right, what you are saying is that `affine` is a world to world transform mapping from:

* domain: the output of applying the `domain_grid2world` to the coordinates implied by `domain_grid_shape`;

to 

* codomain: the output of applying the `codomain_grid2world` to the coordinates implied by `codomain_grid_shape`.

Is that right?